src/Compilers/Core/Portable/DiagnosticAnalyzer/SymbolDeclaredCompilationEvent.cs

```diff
@@ -35,7 +33,18 @@ internal sealed class SymbolDeclaredCompilationEvent : CompilationEvent
         public SemanticModel? SemanticModelWithCachedBoundNodes { get; }
 
         // PERF: We avoid allocations in re-computing syntax references for declared symbol during event processing by caching them directly on this member.
-        public ImmutableArray<SyntaxReference> DeclaringSyntaxReferences => _lazyCachedDeclaringReferences.Value;
+        public ImmutableArray<SyntaxReference> DeclaringSyntaxReferences
+        {
+            get
+            {
+                if (!_lazyCachedDeclaringReferences.HasValue)
+                {
+                    _lazyCachedDeclaringReferences = Symbol.DeclaringSyntaxReferences;
```

Comment:
>_lazyCachedDeclaringReferences = Symbol.DeclaringSyntaxReferences; [](http://example.com/codeflow?start=20&length=66)

This assignment is not atomic. Why we are using a nullable type at all. Is using ```default(ImmutableArray<SyntaxReference>)``` as an initial value and using interlocked initialize going to work?

Reply:
@ToddGrun Specifically, you're going to want to use `InterlockedOperations.Initialize` here with a static lambda callback.

Reply:
I believe the full form would be:

```cs
return InterlockedOperations.Initialize(
  ref _lazyCachedDeclaringReferences,
  static symbol => symbol.DeclaringSyntaxReferences,
  Symbol);
```

Reply:
OK, I've gone ahead with your and Sam's suggestion. I did consider this originally, but wasn't sure whether the Symbol.DeclaringSyntaxReferences could return a default immutable array. But, if that's not a concern, then this is definitely better.

Reply:
Even if it could, we would still need to use some kind of sentinel array to compare with. `Nullable<T>` cannot be used in this scenario, period; you could have two threads race and observe a sheared `Nullable<ImmutableArray>` value in thread 2, such that the write to `Nullable.hasValue` is true, but `Nullable.value` has not been written yet.

Reply:
As I wrote in the original description, I didn't consider multiple assignments during periods of high contention as being a deal breaker. However, I'm more than glad to have made the change such that the value won't be calculated twice as that seems to be the desired consensus.

Reply:
Fred stopped by and explained what I was missing. Not the first time a torn-write has slipped past me.